src/main/kotlin/com/sourcegraph/cody/chat/AgentChatSession.kt

```diff
@@ -339,48 +324,54 @@ private constructor(
+    fun restoreChatSession(
+        agent: CodyAgent,
+        chatState: ChatState
+    ): CompletableFuture<ConnectionId> {
+      val modelFromState =
+          chatState.llm?.let {
+            ChatModelsResponse.ChatModelProvider(
+                default = it.model == null,
+                codyProOnly = false,
+                provider = it.provider,
+                title = it.title,
+                model = it.model ?: "")
+          }
+
+      val chatMessages =
+          chatState.messages.map { message ->
+            val parsed =
+                when (val speaker = message.speaker) {
+                  MessageState.SpeakerState.HUMAN -> Speaker.HUMAN
+                  MessageState.SpeakerState.ASSISTANT -> Speaker.ASSISTANT
+                  else -> error("unrecognized speaker $speaker")
+                }
+
+            ChatMessage(speaker = parsed, message.text)
+          }
+
+      val messages =
+          chatMessages
+              .dropWhile { it.speaker == Speaker.ASSISTANT }
+              .fold(emptyList<ChatMessage>()) { acc, msg ->
+                if (acc.lastOrNull()?.speaker == msg.speaker) acc else acc.plus(msg)
+              }
+
+      val restoreParams = ChatRestoreParams(modelFromState?.model, messages, chatState.internalId!!)
+      return agent.server.chatRestore(restoreParams)
+    }
+
     fun createFromState(project: Project, state: ChatState): AgentChatSession {
+
       val agentChatSession = CompletableFuture<AgentChatSession>()
       createNewPanel(project) { codyAgent ->
-        val newConnectionId = codyAgent.server.chatNew()
-        val modelFromState =
-            state.llm?.let {
-              ChatModelsResponse.ChatModelProvider(
-                  default = it.model == null,
-                  codyProOnly = false,
-                  provider = it.provider,
-                  title = it.title,
-                  model = it.model ?: "")
-            }
-
-        val chatSession =
-            AgentChatSession(project, newConnectionId, state.internalId!!, modelFromState)
-        val chatMessages =
-            state.messages.map { message ->
-              val parsed =
-                  when (val speaker = message.speaker) {
-                    MessageState.SpeakerState.HUMAN -> Speaker.HUMAN
-                    MessageState.SpeakerState.ASSISTANT -> Speaker.ASSISTANT
-                    else -> error("unrecognized speaker $speaker")
-                  }
-
-              ChatMessage(speaker = parsed, message.text)
-            }
-        chatSession.messages.addAll(chatMessages)
-        ApplicationManager.getApplication().invokeLater {
-          chatSession.chatPanel.addAllMessages(chatMessages)
-        }
-
-        if (modelFromState?.model.isNullOrEmpty()) {
-          chatSession.restoreAgentSession(codyAgent)
-        } else {
-          chatSession.restoreAgentSession(codyAgent, modelFromState)
-        }
+        val dummyConnectionId = codyAgent.server.chatNew()
```

Comment:
Why it's `dummyConnectionId` ant not just `connectionId`?

Reply:
this is not the one we use eventually. pls take a look:

```
fun updateFromState(agent: CodyAgent, state: ChatState) {
    val newConnectionId = restoreChatSession(agent, state)
    connectionId.getAndSet(newConnectionId)
    fetchLlms()
  }
```